### PR TITLE
test first version of plotting.boxplot

### PR DESCRIPTION
**Pull request by byersiiasa**
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added (to plotting examples)
- [x] Description in RELEASE_NOTES.md Added
# Description of PR

This PR adds possibility of plotting directly to a boxplot, currently implemented using seaborn. Seaborn implementation is preferred over standard matplotlib/pandas boxplot options due to the ease of adding 3rd dimension "hue" - which allows grouping of boxplots.

### CODE REVIEW

**Comment by byersiiasa** [comment]
@danielhuppmann  can you please check the test failure?
Trying to check the `type` of df in the function - 
`    if type(df) == pyam.core.IamDataFrame:`
but the tests don't like that.

File: pyam/core.py [hunk 1]
```diff
@@ -1531,7 +1531,16 @@ def bar_plot(self, *args, **kwargs):
         df = self.as_pandas(with_metadata=True)
         ax = plotting.bar_plot(df, *args, **kwargs)
         return ax
+        
+    def boxplot(self, *args, **kwargs):
```
Comment (gidden) [inline pyam/core.py:1535]: can this be `box_plot` to match other functions please?

File: pyam/plotting.py [hunk 2]
```diff
@@ -496,6 +497,67 @@ def bar_plot(df, x='year', y='value', bars='variable',
     return ax
 
 
+def boxplot(df, y='value', x='year', hue=None, palette='Spectral_r',
```
Comment (gidden) [inline pyam/plotting.py:500]: same request here - `box_plot()`

File: pyam/plotting.py [hunk 3]
```diff
@@ -496,6 +497,67 @@ def bar_plot(df, x='year', y='value', bars='variable',
...
+            ax=None, legend=False, **kwargs):
+    """ Plot boxplot of data using seaborn.boxplot
+
+    Parameters
+    ----------
+    df : pandas.DataFrame or pyam.IamDataFrame
```
Comment (gidden) [inline pyam/plotting.py:506]: to match other functions, this should only take a `pd.DataFrame` 

the thought here is that these utilities are purely for plotting, and internal manipulations to the `pyam.DataFrame` should happen in that scope (`core.py` or other) and not here

File: pyam/plotting.py [hunk 4]
```diff
@@ -496,6 +497,67 @@ def bar_plot(df, x='year', y='value', bars='variable',
...
+        default: 'Spectral_r'
+    ax : matplotlib.Axes, optional
+    legend : bool, optional
+        Include a legend
+        default: False
+    kwargs : Additional arguments to pass to the pd.DataFrame.plot()
+    """
+
+    # Check data_type
+    if type(df) == pyam.core.IamDataFrame:
```
Comment (gidden) [inline pyam/plotting.py:533]: all data preparation should go in the `box_plot()` function in `core.py`. see, e.g., https://github.com/IAMconsortium/pyam/blob/90bcef21019217d7030a5d94d8d142ad3fbdc6af/pyam/core.py#L1497

**Comment by gidden** [comment]
cool thanks @byersiiasa ! it looks like this only works for variables at the moment, do you want to extend to other data (e.g., metadata, model, scenario)?

File: pyam/core.py (hunk 1 above)
Comment (danielhuppmann) [reply pyam/core.py:1535]: I disagree, I would follow the pandas convention where it's also called `boxplot`

File: pyam/plotting.py (hunk 2 above)
Comment (danielhuppmann) [reply pyam/plotting.py:500]: as above, I'd rather keep it as is...

**Comment by byersiiasa** [comment]
> cool thanks @byersiiasa ! it looks like this *only works for variables* at the moment, do you want to extend to other data (e.g., metadata, model, scenario)?

@gidden   I think `y` probably only  `value` or  a numerical column from meta_data; but `x` and `hue` should be fine taking `model`, `scenario`, region` , `year` or anything else.

Or do you mean something else?

File: pyam/plotting.py (hunk 3 above)
Comment (danielhuppmann) [reply pyam/plotting.py:506]: agree with @gidden - let's not overload the functionality within the plotting module, this should only receive correctly formatted data from `IamDataFrame.boxplot()` or a similarly formatted pandas.DataFrame

File: pyam/plotting.py [hunk 5]
```diff
@@ -496,6 +497,62 @@ def bar_plot(df, x='year', y='value', bars='variable',
...
+        Include a legend
+        default: False
+    kwargs : Additional arguments to pass to the pd.DataFrame.plot() 
+    
+    
+    """
+    
+    # Check data_type
+    if type(df) == pyam.core.IamDataFrame:
+        if hue==None:
```
[the remark this answers is not in the document]
Comment (danielhuppmann) [reply pyam/plotting.py:534]: this could be just `if not hue`

File: RELEASE_NOTES.md [hunk 6]
```diff
@@ -1,5 +1,5 @@
 # Next Release
-
+- [#395] (https://github.com/IAMconsortium/pyam/pull/403_ - Add boxplot function
```
Comment (danielhuppmann) [inline RELEASE_NOTES.md:2]:
```suggestion

- [#403](https://github.com/IAMconsortium/pyam/pull/403) Add boxplot function.
```

**Comment by danielhuppmann** [comment]
> @danielhuppmann can you please check the test failure?
> Trying to check the `type` of df in the function -
> ` if type(df) == pyam.core.IamDataFrame:`
> but the tests don't like that.

Just tried running this is a notebook, it also doesn't like it. In general, you want `isinstance(df, IamDataFrame)`, but in this case (as noted by @gidden) the data-processing should be moved out of this function anyway...

**Comment by danielhuppmann** [comment]
Thanks @byersiiasa for taking a first stab at a boxplot function!

I'm a bit worried that having plots created by different libraries will make maintenance more difficult going forward... But I also see (after some playing around now) that seaborn is easier than matplotlib for this.

Still, two recommendations:
 - follow the pandas style and use `by` rather than `hue` in the public-facing API - after all, the aim is to group the data *by*. some column(s)...
 - it would be nice to being able to use the colors set in runcontrol out-of-the-box

**Comment by byersiiasa** [comment]
thanks @gidden @danielhuppmann 
Ok. Implemented almost all suggested changes, e.g.
`by` instead of `hue`
made more consistent with other plotting  functions
have left as `boxplot` for now - you will have to duel it, but my thinking was same as Daniels.

Comment about `run_control` is acknowledged, may try another time. Have removed the `palette` argument for now - user can specify as they want.

Tested the `plot_boxplot` example script - but the data is not hugely suited for boxplot. ideally would have more models or scenarios. This could be added to `data.csv`, with caveat that `filter` calls in every plotting example may need to be checked / adjusted
or could add a separate dataset e.g. `data_bp.csv`

**Comment by danielhuppmann** [comment]
> Tested the plot_boxplot example script - but the data is not hugely suited for boxplot. ideally would have more models or scenarios.

I would be in favour of a better dataset for the plot examples (and tests)...

**Comment by byersiiasa** [comment]
ok - so hopefully this has nice enough example plot for @danielhuppmann to have a look
I added some more scenarios to the `-examples/data.csv` - hope that's ok!

**Comment by danielhuppmann** [comment]
Thanks @byersiiasa - as discussed, I squash-and-rebased your branch to the latest master, and started a new PR.
Closing in favor of #440

**Pull request closed without merging**